### PR TITLE
feat(treeview): add treeview component

### LINKED ISSUE #3206: Introduce Tree menu
Introduce a Tree menu component per the designs outlined here: https://marvelapp.com/8cdg211/screen/68891042

### PR DESCRIPTION
**Pull request by christiemolloy**
closes #3206 

new PR from new fork

old PR #3303

### CODE REVIEW

File: src/patternfly/components/TreeView/tree-view.scss [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+  // Link
+  --pf-c-tree-view__link--PaddingTop: var(--pf-global--spacer--sm);
+  --pf-c-tree-view__link--PaddingRight: var(--pf-global--spacer--sm);
+  --pf-c-tree-view__link--PaddingBottom: var(--pf-global--spacer--sm);
+  --pf-c-tree-view__link--PaddingLeft: var(--pf-global--spacer--sm);
+  --pf-c-tree-view__link--Color: var(--pf-global--Color--100);
+  --pf-c-tree-view__link--m-current--Color: var(--pf-global--link--Color);
+  --pf-c-tree-view__link--m-current--FontWeight: var(--pf-global--FontWeight--bold);
+  --pf-c-tree-view__link--m-first--FontWeight: var(--pf-global--FontWeight--bold);
```
Comment (mattnolting) [inline src/patternfly/components/TreeView/tree-view.scss:20]: what do you think about `m-root` instead of `m-first`.

Comment (christiemolloy) [reply src/patternfly/components/TreeView/tree-view.scss:20]: yep thats good

Comment (christiemolloy) [reply src/patternfly/components/TreeView/tree-view.scss:20]: yeah thats a good idea!

File: src/patternfly/components/TreeView/tree-view.scss [hunk 2]
```diff
@@ -0,0 +1,185 @@
+$pf-c-tree-view--MaxNesting: 10;
+
+.pf-c-tree-view {
+  --pf-c-tree-view--PaddingTop: var(--pf-global--spacer--sm);
+  --pf-c-tree-view--PaddingBottom: var(--pf-global--spacer--sm);
```
Comment (mcoker) [inline src/patternfly/components/TreeView/tree-view.scss:5]: do we need this or should it come from the parent layout? This padding exists in the `pf-c-[dropdown-menus]__menu element`, which is where tree view would go, so you would have double padding.

Comment (christiemolloy) [reply src/patternfly/components/TreeView/tree-view.scss:5]: maybe a padding modifier to achieve this padding ?

File: src/patternfly/components/TreeView/tree-view.scss [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  align-items: center;
+  justify-content: center;
+  min-width: var(--pf-c-tree-view__link-toggle-icon--MinWidth);
+  min-height: var(--pf-c-tree-view__link-toggle-icon--MinHeight);
+  padding-right: var(--pf-c-tree-view__link-toggle-icon--PaddingRight);
+}
+
+.pf-c-tree-view__text {
+  @include pf-text-overflow;
+  @include pf-line-clamp("var(--pf-c-tree-view__text--max-lines)");
```
Comment (mcoker) [inline src/patternfly/components/TreeView/tree-view.scss:129]: I think we just want to use `@include pf-text-overflow` here?

File: src/patternfly/components/TreeView/examples/TreeView.md [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+| -- | -- | -- |
+| `.pf-c-tree-view` | `<div>` | Initiates the tree view container. **Required** |
+| `.pf-c-tree-view__list` | `<ul>` | Initiates a tree view list. **Required** |
+| `.pf-c-tree-view__list-item` | `<li>` | Initiates a tree view list item. **Required** |
+| `.pf-c-tree-view__content` | `<div>` | Initiates a tree view node. **Required** |
+| `.pf-c-tree-view__link` | `<button>`, `<a>` | Initiates a tree view link. **Required** |
+| `.pf-c-tree-view__link-count` | `<span>` | Initiates a tree view link count. |
+| `.pf-c-tree-view__toggle-icon` | `<span>` | Initiates a tree view toggle icon. |
+| `.pf-c-tree-view__text` | `<span>` | Initiates tree view text. |
+| `.pf-c-tree-view__icon` | `<span>` | Initiates a tree view icon. |
```
Comment (mcoker) [inline src/patternfly/components/TreeView/examples/TreeView.md:449]:
```suggestion
| `.pf-c-tree-view__link-icon` | `<span>` | Initiates a tree view icon. |
```

File: src/patternfly/components/TreeView/examples/TreeView.md [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+### Usage
+
+| Class | Applied | Outcome |
+| -- | -- | -- |
+| `.pf-c-tree-view` | `<div>` | Initiates the tree view container. **Required** |
+| `.pf-c-tree-view__list` | `<ul>` | Initiates a tree view list. **Required** |
+| `.pf-c-tree-view__list-item` | `<li>` | Initiates a tree view list item. **Required** |
+| `.pf-c-tree-view__content` | `<div>` | Initiates a tree view node. **Required** |
+| `.pf-c-tree-view__link` | `<button>`, `<a>` | Initiates a tree view link. **Required** |
+| `.pf-c-tree-view__link-count` | `<span>` | Initiates a tree view link count. |
```
Comment (mcoker) [inline src/patternfly/components/TreeView/examples/TreeView.md:446]: need to add `pf-c-tree-view__link-check`

File: src/patternfly/components/TreeView/examples/TreeView.md [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+| `.pf-c-tree-view__toggle-icon` | `<span>` | Initiates a tree view toggle icon. |
```
Comment (mcoker) [inline src/patternfly/components/TreeView/examples/TreeView.md:447]:
```suggestion
| `.pf-c-tree-view__link-toggle-icon` | `<span>` | Initiates a tree view toggle icon. |
```

File: src/patternfly/components/TreeView/tree-view-link-toggle.hbs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<span class="pf-c-tree-view__link-toggle-icon{{#if tree-view-link-toggle-icon--modifier}} {{tree-view-link-toggle-icon--modifier}}{{/if}}"
```
Comment (mcoker) [inline src/patternfly/components/TreeView/tree-view-link-toggle.hbs:1]: this file should be `tree-view-link-toggle-icon.hbs`

**Comment by mturley** [comment]
@mcoker @christiemolloy is there already someone assigned to implement this for patternfly-react? I'm not finding an issue there. I need to implement this in React downstream, and I think I may contribute it back to PF if nobody else is working on it.

**Comment by mturley** [comment]
Also, in the example I see that most of the `<li>` elements have `tabindex="0"` but some have `tabindex="-1"`. When should one be used versus the other?

**Comment by mturley** [comment]
I started a React implementation of the basic tree view use case here: https://github.com/konveyor/mig-ui/pull/985
If nobody else beats me to it, I plan to implement the rest of the features and contribute this to pf-react. cc @evwilkin

**Comment by mcoker** [comment]
Hey @mturley here's the patternfly-react issue - https://github.com/patternfly/patternfly-react/issues/4652

**Comment by mturley** [comment]
Awesome, thanks @mcoker !

**Pull request merged**